### ISSUE TITLE
plugin/forward:  maxes out CPU (TLS connection negotiating)

### ISSUE THREAD

**Issue by WinstonPrivacy** [coredns/coredns#2624]
We've noticed an occasional problem with using CoreDNS on a low-end ARM64 board (1Gb RAM, dual core 1.0 Ghz CPU). 

If another process temporarily maxes out CPU, CoreDNS CPU spikes to 80%+ and stays there indefinitely, even after the other process has been terminated. 

The issue only appears if there is a lot of DNS activity during the period of peak CPU activity. My guess is that the DNS queries pile up but CoreDNS is unable to ever catch up with them for some reason. As an example, here is a screenshot of CoreDNS log output taken a few minutes after terminating a high-CPU process. This will continue indefinitely and in fact, the response times almost seem to get worse.

The problem goes away once CoreDNS is restarted. Happy to try and help diagnose this but I'm not sure where to start.

![coredns failure](https://user-images.githubusercontent.com/36650369/53460250-15b46400-3a02-11e9-9e4b-40756c70221b.png)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] CoreDNS maxes out..." ]
>We've noticed an occasional problem with using CoreDNS on a low-end ARM64 board (1Gb RAM, dual core 1.0 Ghz CPU).
>
>If another process temporarily maxes out CPU, CoreDNS CPU spikes to 80%+ and stays there indefinitely, even after the other process has been terminated.
>
>The issue only appears if there is a lot of DNS activity during the period of peak CPU activity. My guess is that the DNS queries pile up but CoreDNS is unable to ever catch up with them for some reason. As an example, here is a screenshot of CoreDNS log output taken a few minutes after terminating a high-CPU process. This will continue indefinitely and in fact, the response times almost seem to get worse.
>
>The problem goes away once CoreDNS is restarted. Happy to try and help diagnose this but I'm not sure where to start.
>
>![coredns failure](https://user-images.githubusercontent.com/36650369/53460250-15b46400-3a02-11e9-9e4b-40756c70221b.png)

What version are you running and with what Corefile?

Do you have some metrics (enable prometheus plugin), like number of goroutines, memory and cpu usage. A
pprof (enable pprof plugin) could help as well.

**Comment by johnbelamaric** [comment]
At first glance this tells me we need to put a limit on our inbound request
queue and start dropping packets. There was another issue that seemed to
imply the same thing.

**Comment by miekg** [comment]
Agreed. Except is really hard to decide when to drop. Also the end user
result is pretty bad: no resolving.

My guess is probably the same as yours, but the requests prolly pile up in
the worker goroutines we spawn. Having easy access to when we received them
and discarding them after, like, 6 seconds seems prudent, as the client as
long given up on them.

**Comment by WinstonPrivacy** [comment]
A configurable threshold limit for inbound requests would be excellent. Reaching it could trigger some kind of clean up logic. For instance, in this catastrophic case, many of the requests are duplicated. Just resolving them all at once might help a great deal. 

Dropping 80% of the queue might is another simplistic but effective solution. In our case, it really doesn't matter because everything is resolved with SERVFAIL, so nothing is getting through. Dropping the entire queue would even work.

A LIFO approach would be good (oldest requests get dropped first because the client is more likely to have given up on them).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS maxes..." ]
>A configurable threshold limit for inbound requests would be excellent. Reaching it could trigger some kind of clean up logic. For instance, in this catastrophic case, many of the requests are duplicated. Just resolving them all at once might help a great deal.
>
>Dropping 80% of the queue might is another simplistic but effective solution. In our case, it really doesn't matter because everything is resolved with SERVFAIL, so nothing is getting through. Dropping the entire queue would even work.
>
>A LIFO approach would be good (oldest requests get dropped first because the client is more likely to have given up on them).

This is all hard to get right.

Mean while a repro would imply getting a coredns process too few CPU cycles and 
bombard it with queries and see what happens.

If you can easily repro in your setup a cpu pprof and a prometheus screenshot 
would help.

Thanks

**Comment by willwp** [comment]
I have repro'd on coredns, same build, same platform as @WinstonPrivacy : spike coredns CPU, then bombard coredns with queries.  Attached are prometheus screen shots.  Notice that go_routines spike to > 2000 when reproducing.   I cannot fetch pprof data when enabled in the Corefile when reproducing the issue.

![go_routines](https://user-images.githubusercontent.com/40573407/53690765-6c89a880-3d36-11e9-944d-e55b029dcefb.png)

![process_cpu_seconds_total](https://user-images.githubusercontent.com/40573407/53690767-73182000-3d36-11e9-93e2-132758df26fd.png)

**Comment by fturib** [comment]
- for the go routines, I think that is expected: there is no limitation of the go routine to process a DNS msg. The limitation is **only** on keeping that routine alive to process subsequent msg

see : https://github.com/miekg/dns/blob/092d7745b47c1a19a6533514ff7a05387b5c5bf9/server.go#L244-L255

I agree that anyway, we should somehow limit the number of go routine running in parallel and ensure we do not go higher than a reasonable amount of memory.

- What is concerning is that we do not release the memory after the spike of msgs. Looks like that once the system is stressed, it keeps it's memory.

@willwp : can you share the Corefile for this test, and the version of the image used ?

**Comment by willwp** [comment]
image:  CoreDNS-1.2.5, linux/arm64, go1.11.2 , with this fix merged in: https://github.com/coredns/coredns/pull/2319
Corefile
~~~~
:53 {
       cache 3600
       forward . tls://1.1.1.1 tls://1.0.0.1 {
               tls_servername cloudflare-dns.com
               health_check 5s
               max_fails 1
               expire 90s
       }
       alternate SERVFAIL . tls://9.9.9.9 tls://149.112.112.112 {
                tls_servername dns.quad9.net
                health_check 5s
                max_fails 1
                expire 90s
       }
}
.:54 {
       cache 3600
       forward . tls://1.1.1.1 tls://1.0.0.1 {
               tls_servername cloudflare-dns.com
               health_check 5s
               max_fails 1
               expire 90s
        }
        alternate SERVFAIL . tls://9.9.9.9 tls://149.112.112.112 {
               tls_servername dns.quad9.net
               health_check 5s
               max_fails 1
               expire 90s
        }
}
~~~~

**Comment by fturib** [comment]
@willwp : the case your propose does not reproduce the same case as @WinstonPrivacy - IMO.

He is describing high CPU utilisation after a spike of queries, like if CoreDNS could never catch-up with the incoming queries and CoreDNS returning Servfail with long time queries.

In your case, you have a spike of incoming queries, the number of go routine is going high (2k), but after the peak, the go routines are returning to normal, along with the CPU (we can see the slope of process_cpu-second going flat), and the the total request processed.

You are showing an issue I think : I do not see why the memory used stay that high after the burst of queries


=> **Can you opened a separated issue for this case** ? (I mean keep separated the high CPU of this issue vs the high amount of memory of your case).

**Comment by willwp** [comment]
@fturib I'm on the same team as @WinstonPrivacy and followed the same steps to reproduce, on the same hardware/software stack.  CoreDNS gets in the high CPU utilization after a spike of queries, when another process temporarily maxes out CPU, in my case coredns maxing out CPU is occurring.  CPU maxing out at > 80%-100% and staying there.  I cannot fetch with pprof from coredns when this CPU spike occurs.  Is there a recommended method for me to properly collect pprof data?  Are there other stats that would be helpful from Prometheus?

**Comment by fturib** [comment]
hum ... if you max out the CPU and stay there, why the graph of `process_cpu_second` is going flat after the burst ? For me it means we are not consuming the CPU anymore (almost) after the burst ...

When you say `CoreDNS is maxing out CPU and stay there` : what tool provide you the CPU used by CoreDNS ?

What is your environment here  ?  Are you in a Kubernetes cluster ? or CoreDNS is running as a process ?

**Comment by willwp** [comment]
I'm running same platform as @WinstonPrivacy low-end ARM64 board (1Gb RAM, dual core 1.0 Ghz CPU), CoreDNS running as a process.    I run htop during reproducing this problem and see CPU for coredns pegged > 80%.  Since pprof was not able to fetch any data, is it possible that the Prometheus client/plugin was not able to provide accurate CPU data for the Prometheus server?

**Comment by fturib** [comment]
> I run htop during reproducing this problem and see CPU for coredns pegged > 80%.

- You are not in a container environment ? (Kubernetes ? or Docker directly ?)
- You are running a CoreDNS from cmd line of linux ? (or a linux service ?)

> Since pprof was not able to fetch any data, is it possible that the Prometheus client/plugin was not able to provide accurate CPU data for the Prometheus server?

The data seems very nice .. it is not random points. 
- Can it happen that Prometheus here is pointing another instance of CoreDNS ? (I guess not, because we can see the burst).
- Can it happen that the `htop coredns` is monitoring something bigger than ONLY CoreDNS ?

**Comment by willwp** [comment]
>     * You are not in a container environment ? (Kubernetes ? or Docker directly ?)
No I'm not in a container environment.
>     * You are running a CoreDNS from cmd line of linux ? (or a linux service ?)
Running CoreDNS as a daemon, linux service, /etc/init.d/coredns
 
> The data seems very nice .. it is not random points.
> 
>     * Can it happen that Prometheus here is pointing another instance of CoreDNS ? (I guess not, because we can see the burst).
It is not pointing to another instance of CoreDNS.  There are no other instances of CoreDNS to monitor in my network or setup. There are no other Prometheus clients running on this system, only the CoreDNS plugin.  
>     * Can it happen that the `htop coredns` is monitoring something bigger than ONLY CoreDNS ?
htop displays the CPU usage of each process running on the system, and htop shows that the CoreDNS process is utilizing > 80%.

**Comment by SmashGuy2** [comment]
The issue is easy to reproduce. Is there logging code we can add to narrow down what's happening?

**Comment by fturib** [comment]
Sorry guys, I was busy on several other projects, and on the memory issue #2593.

But now re-reading and after the test @rajansandeep did for that memory issue.
I am thinking that if it is easy to reproduce, the best would be:

1- add the `pprof` plugin to the corefile
2- reproduce the mode with permanent high CPU (80%). I mean after the burst, when the go routines are back to almost zero.
3- activate collection of data pprof for the cpu
4- look at the result : we should see immediately in which proc the CPU is spent and most likely understand the problem.

Would that be feasible ?

**Comment by WinstonPrivacy** [comment]
The pprof plugin apparently isn't working when we reproduce this. We're going to rebuild and manually try to generate a pprof cpuprofile.

**Comment by willwp** [comment]
@fturib :  Is this outlined [solution](https://github.com/coredns/coredns/issues/2593#issuecomment-470307464) in #2593 predicted to address what we're seeing in this ticket?

**Comment by willwp** [comment]
@fturib:  I followed the proposed change by @tmthrgd in his [comment](https://github.com/coredns/coredns/issues/2593#issuecomment-470323104) .   
Improvements noticed:

- CoreDNS recovered from > 80% CPU after 2-5 minutes, after the other resource hungry process is done using CPU and memory resources.

-  CoreDNS was no longer stuck in SERVFAIL, and within 2-5 minutes starts resolving DNS queries.

Before @tmthrgd's change it would take CoreDNS > 2 hours to recover, even after other resource hungry processes were stopped.

**Comment by fturib** [comment]
@willwp : I do not relate #2593 with the description of this issue, although both are facing a problem of behavior after a burst of query.

- in #2593 : we do not have the CPU high after the query burst, but the memory is an issue **during** the burst.
- in current one, you are facing a memory spike that does not resolve after the burst. And you are facing also high CPU **after*8 the burst, I mean after the go-routines are going down.

The difference can comes from difference of environment and of and maybe test:
* Kubernetes pod for #2593 against direct linux process in current issue
* test: #2593 is querying always the same 10 or 100 fqdn ... **what about your case** ?

for the configuration (Corefile), it is different but in both case, the main part of the query processing  is just a cache hit. 


I am happy to see that proposition with no workers is working for you ... but I have no idea why. 
I do not understand yet where the CPU is being used in your case.

**Comment by fturib** [comment]
@willwp : I am curious to see the description of the kind of stress you are exercising during your test. 
I mean what is the QPS, and if the domain queried are always the same (in that case we hit the `cache` most often) or very diverse (in that case we hit more ofthe the `forward` plugin)

**Comment by willwp** [comment]
@fturib :  Here is how I run the test, and the domains are diverse. 

- Run coredns command line with log enabled ./coredns -conf Corefile

- dnsperf -s x.x.x.x -p 53 -d queryfile-example-10million-201202 -l 300 -c 30 -Q 2000

- Run some devices mobile, streaming, apps that make a lot of queries

-  On the device build a substantial go program, such as coredns, to stress the memory and CPU.

I plan on removing the change from #2593 and reproducing with pprof compiled in coredns.go, because I cannot fetch anything from the pprof plugin during the failure.

coredns.go enabling cpu profiling.
~~~~
package main

//go:generate go run directives_generate.go
import (
        "github.com/coredns/coredns/coremain"

        // Plug in CoreDNS
        _ "github.com/coredns/coredns/core/plugin"
        // Profiling, remove
      "github.com/pkg/profile"
)

func main() {
        // CPU profiling
       defer profile.Start().Stop()
        coremain.Run()
}
~~~~

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS maxes..." ]
>@willwp : I am curious to see the description of the kind of stress you are exercising during your test.
>I mean what is the QPS, and if the domain queried are always the same (in that case we hit the `cache` most often) or very diverse (in that case we hit more ofthe the `forward` plugin)

this is the same problem as #2593 as root-caused by @fturb. You're on this thread
https://github.com/miekg/dns/issues/916 as well (I think) for a proper fix.

**Comment by WinstonPrivacy** [comment]
So we've been testing the latest and the problem is still present. If resources become limited, CoreDNS maxes out CPU and becomes unresponsive for 5-30 minutes (better than before but still catastrophic). What's making this worse is that CoreDNS then starves out other processes, which causes one of them to do something similar. On one install, this is then triggering a smart TV to make thousands of repeated requests, which makes the problem so bad that all services need to be manually restarted (or the box rebooted).

**Comment by fturib** [comment]
@WinstonPrivacy , @willwp : can you re-run your test with the [enhanced `pprof`](https://github.com/coredns/coredns/pull/2729). and collect the pprof with the `block` option so we can see what these goroutines are waiting for.

if your test [with `throttle`](https://github.com/coredns/coredns/pull/2706#issuecomment-476795457) is corret, then you have goroutines spinning up whatever the limitation by throttle. 

the CPU profiling shows that most of the time is spent in TLS (decrypting the handshake). And that handshake process is spinning up a goroutine ... That would explain the number of goroutines, the memory and also the CPU.

To be sure we need to see where the goroutines of the test are stucked ... (like we did to validate the [usage of throttle](https://github.com/coredns/coredns/issues/2593#issuecomment-476363778)).

I propose you to use the same corefile, extending pprof options:

```
.:53 {                                                       
       cache 3600                                        
       forward . tls://1.1.1.1 tls://1.0.0.1 {
               tls_servername cloudflare-dns.com
               health_check 5s                  
               max_fails 1                      
               expire 90s                       
       }                                        
       alternate SERVFAIL . tls://9.9.9.9 tls://149.112.112.112 {
                tls_servername dns.quad9.net                     
                health_check 5s                                  
                max_fails 1                                      
                expire 90s                                       
       }                                    
       throttle 1000              
       prometheus 0.0.0.0:9090
       pprof 0.0.0.0:6053  1
}    
```                         

- collect pprof for cpu and block.
- and for on premetheus, you can monitor the variable : rate(coredns_throttle_seen_queries, [1m]), and coredns_throttle_flight_queries as simple value.

**Comment by willwp** [comment]
> @WinstonPrivacy , @willwp : can you re-run your test with the [enhanced `pprof`](https://github.com/coredns/coredns/pull/2729). and collect the pprof with the `block` option so we can see what these goroutines are waiting for.

Yes see results below

> I propose you to use the same corefile, extending pprof options:

We used the proposed Corefile

>     * collect pprof for cpu and block.
* pprof cpu
```
(pprof) top
Showing nodes accounting for 28.25s, 52.36% of 53.95s total
Dropped 555 nodes (cum <= 0.27s)
Showing top 10 nodes out of 148
      flat  flat%   sum%        cum   cum%
     4.47s  8.29%  8.29%      5.12s  9.49%  runtime.step /usr/local/go/src/runtime/symtab.go
     4.27s  7.91% 16.20%     24.49s 45.39%  math/big.nat.divLarge /usr/local/go/src/math/big/nat.go
     3.09s  5.73% 21.93%      3.71s  6.88%  runtime.findObject /usr/local/go/src/runtime/mbitmap.go
     3.06s  5.67% 27.60%      3.06s  5.67%  math/big.mulAddVWW /usr/local/go/src/math/big/arith_arm64.s
     2.89s  5.36% 32.96%      2.89s  5.36%  math/big.addMulVVW /usr/local/go/src/math/big/arith_arm64.s
     2.61s  4.84% 37.79%      7.85s 14.55%  runtime.pcvalue /usr/local/go/src/runtime/symtab.go
     2.42s  4.49% 42.28%      2.49s  4.62%  math/big.divWW_g /usr/local/go/src/math/big/arith.go
     2.23s  4.13% 46.41%      7.32s 13.57%  runtime.scanobject /usr/local/go/src/runtime/mgcmark.go
     1.84s  3.41% 49.82%     13.56s 25.13%  runtime.mallocgc /usr/local/go/src/runtime/malloc.go
     1.37s  2.54% 52.36%     13.06s 24.21%  runtime.gentraceback /usr/local/go/src/runtime/traceback.go
```
[pprof_cpu.svg.zip](https://github.com/coredns/coredns/files/3011752/pprof_cpu.svg.zip)


* pprof block
```
(pprof) top
Showing nodes accounting for 7.10days, 99.92% of 7.10days total
Dropped 99 nodes (cum <= 0.04days)
Showing top 10 nodes out of 53
      flat  flat%   sum%        cum   cum%
  5.50days 77.42% 77.42%   5.50days 77.42%  sync.(*Mutex).Lock /usr/local/go/src/sync/mutex.go
  1.54days 21.69% 99.11%   1.54days 21.69%  runtime.chanrecv1 /usr/local/go/src/runtime/chan.go
  0.06days  0.81% 99.92%   0.06days  0.81%  runtime.selectgo /usr/local/go/src/runtime/select.go
         0     0% 99.92%   4.95days 69.66%  crypto/ecdsa.Verify /usr/local/go/src/crypto/ecdsa/ecdsa.go
         0     0% 99.92%   0.54days  7.61%  crypto/elliptic.(*CurveParams).IsOnCurve /usr/local/go/src/crypto/elliptic/elliptic.go
         0     0% 99.92%   3.67days 51.65%  crypto/elliptic.(*CurveParams).ScalarBaseMult /usr/local/go/src/crypto/elliptic/elliptic.go
         0     0% 99.92%   4.74days 66.72%  crypto/elliptic.(*CurveParams).ScalarMult /usr/local/go/src/crypto/elliptic/elliptic.go
         0     0% 99.92%   1.42days 20.05%  crypto/elliptic.(*CurveParams).addJacobian /usr/local/go/src/crypto/elliptic/elliptic.go
         0     0% 99.92%   3.31days 46.65%  crypto/elliptic.(*CurveParams).doubleJacobian /usr/local/go/src/crypto/elliptic/elliptic.go
         0     0% 99.92%   0.54days  7.61%  crypto/elliptic.Unmarshal /usr/local/go/src/crypto/elliptic/elliptic.go
``` 
[pprof_block.svg.zip](https://github.com/coredns/coredns/files/3011753/pprof_block.svg.zip)


>     * and for on premetheus, you can monitor the variable : rate(coredns_throttle_seen_queries, [1m]), and coredns_throttle_flight_queries as simple value.

Was not able to get coredns_throttle_flight_queries to graph:

![coredns_throttle_seen_queries](https://user-images.githubusercontent.com/40573407/55053987-20c3d980-502c-11e9-859d-2b479d6af58f.png)



![go_routines_block](https://user-images.githubusercontent.com/40573407/55054017-3fc26b80-502c-11e9-8664-d07bffc13893.png)
![dns_count_block](https://user-images.githubusercontent.com/40573407/55054023-43ee8900-502c-11e9-87e0-27b55362b32b.png)

**Comment by fturib** [comment]
- Looks like you are not reaching the limit of in-flight queries defined on `throttle`.

However, you have the number of goroutine that is going up to 10k (and the mem up to 800Mb).

Because of throttle, we can say that these goroutines are not the one that are processing the msg.

Profiles:
- CPU profile shows that the time spent is in "crypto/tls"
- Block profile show that these go-routines are mostly stuck in "cyrpto/tls" clientHandshake(...)", which is consistent with CPU profile. By the way it seems to be on its own goroutine.

There is a path that generate one go-routine each time, that may stay up longer than the ServeDNS routine because of the timeout:

- Forward.ServeDNS
- Proxy.Connect
- Transport.Dial
- dns.DialTimeoutWithTLS
- Client.Dial
- tls.DialWithDialer  <= This one is the 
- go ... conn.Handshake()
-  .. conn.clientHandshake() <= this one is in CPU and Block profiles


Which means for me: the investigation has to dive into Forward/Proxy and tls with Timeout.

**Comment by fturib** [comment]
You say in description that you are on `low-end ARM64 board`.

Can it happen that the TLS encyrption/decryption is very slow and that the DNS request timed-out because it cannot just connect uptream in delay (timeout on the TLS handshake) ?

And the query return SERVFAIL, but the spwan go-routine for the TLS handshake continue to decrypt ... slowly.

**Comment by WinstonPrivacy** [comment]
I have been playing with this for much of the day. Anecdotally, the
"cancel" build seemed to perform best. It is more difficult to trigger the
high CPU condition, though not impossible.

Your theory seems reasonable to me... But shouldn't a timeout kick in at
some point?

**Comment by johnbelamaric** [comment]
So, is there any "parent" go-routine information in the go runtime so that when we timeout and return SERVFAIL, we can stop any go routines spawned by us?

Alternatively, we can track total number of go routines and throttle on that (assuming the overhead of getting that number is low).

**Comment by fturib** [comment]
> So, is there any "parent" go-routine information in the go runtime so that when we timeout and return SERVFAIL, we can stop any go routines spawned by us?

Is there a way to `stop` a go-routine like we `kill` a process ?

**Comment by johnbelamaric** [comment]
This is interesting though. Do we re-use the TLS client in forward across multiple requests? If we timeout after only 5 seconds, and TLS negotiation takes 10 seconds, then unless we complete the negotiation and then re-use that connection in future requests, we will never be able to successfully resolve a query.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS maxes..." ]
>This is interesting though. Do we re-use the TLS client in forward across multiple requests? If we timeout after only 5 seconds, and TLS negotiation takes 10 seconds, then unless we complete the negotiation and then re-use that connection in future requests, we will never be able to successfully resolve a query.

We re-use the tcp/tls *connection*, which is a *dns.Conn, which wraps a 
net.Conn. I'm not sure (now) if this is enough to re-use a connection fully or 
that some other bits will be needed as well (i.e. client state that we through 
away).

The original intent was very much to establish a connection once, re-use a 
connection for up to 10s (can be made longer) and then reconnect.

Btw: nice work root-causing this down to TLS connection negotiating.

**Comment by WinstonPrivacy** [comment]
This makes sense. I've done tests in the past where I've switched from
DNS-over-TLS to ordinary unencrypted DNS and the problem seems to go away.
It's been awhile since I've done this though.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>This makes sense. I've done tests in the past where I've switched from
>DNS-over-TLS to ordinary unencrypted DNS and the problem seems to go away.
>It's been awhile since I've done this though.

What would be interesting is to see if a re-negotiate happens on every query or 
not. An easy way would be to set `expiration` to higher value. Some `printf` 
might be needed to see if a re-dial is happening as well.

**Comment by vielmetti** [comment]
@WinstonPrivacy - what version of Go are you building this with? I seem to recall that there have been several rounds of perf needed to get TLS on arm64 up to speed, dating all the way back to https://github.com/golang/go/issues/19840

also, having TLS spike CPU was noted by gnatsd at https://github.com/nats-io/gnatsd/issues/589

**Comment by willwp** [comment]
We are building with gov1.12.

**Comment by miekg** [comment]
Also another thing I proposed in some other issue was to have 10 connection ready to go (or at least some minimal), this would reduce the problem of not having any and, in essence, the thundering herd problem of establishing a whole bunch when there's a burst.

**Comment by vielmetti** [comment]
@willwp is it TLS 1.2 or TLS 1.3, and do you know which ciphers you are ending up with? I know that anything that depends on ChaCha20Poly1305 will be particularly slow on arm64 until some SIMD code gets merged.

https://github.com/golang/go/issues/22809

**Comment by johnbelamaric** [comment]
@miekg sure, a connection pool would help, but ideally we would still have some way to quash any goroutines spawned by a request when we time it out

**Comment by miekg** [comment]
It's not as easy, because the connections are shared. We *good*, maybe gate the number of concurrent connections being done on GOMAXPROC, i.e. no more than that.

That would mean holding up the connection dialling in `transport.Dial`, where we don't get a connection from the pool, we disallow more than GOMAXPROC as the same time and then use and re-use those.

That still doesn't prevent 1000's of waiting to be served, which means this also needs to tie into something in `persistent.go` to wait for connections to show up.

**Comment by miekg** [comment]
Just looking through that code now:
* I see a `tls.ClientConnectionState` - which we don't have here now - sounds useful
* there is a LRU impl, which implies more locking, which makes me wonder about the feasibility using this for UDP connection.

UDP conn are cached as well in our impl. to avoid a round trip to kernel space to get a new one.

**Comment by WinstonPrivacy** [comment]
You will almost certainly need more than GOMAXPROCs connections, as a
typical web request will make 10-20 (or even 100+) DNS queries. We can
easily handle 50 concurrent connections on 2 cores and have stress tested
as high as 13,000.

I'm not sure that this can be implemented in persistent.go. While you could
enforce a threshold there, doing so would force you to decide between
blocking incoming connection requests (bad) or failing. The caller
(Forward:connect.go:Dial()) then would just try to open another connection,
unless it was smart enough to enforce a threshold.

I do think persistent.go needs a maxIdleConnection setting.

All of this seems fairly straightforward to implement. I need to figure out
how to cross-compile so I can develop against CoreDNS directly and test
these ideas out on our low-end ARM board.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>You will almost certainly need more than GOMAXPROCs connections, as a
>typical web request will make 10-20 (or even 100+) DNS queries. We can
>easily handle 50 concurrent connections on 2 cores and have stress tested
>as high as 13,000.
>
>I'm not sure that this can be implemented in persistent.go. While you could
>enforce a threshold there, doing so would force you to decide between
>blocking incoming connection requests (bad) or failing. The caller
>(Forward:connect.go:Dial()) then would just try to open another connection,
>unless it was smart enough to enforce a threshold.
>
>I do think persistent.go needs a maxIdleConnection setting.
>
>All of this seems fairly straightforward to implement. I need to figure out
>how to cross-compile so I can develop against CoreDNS directly and test
>these ideas out on our low-end ARM board.

Cross compiling is as easy as setting a GOARCH, see the Makefile.release on how 
to do this.

**Comment by WinstonPrivacy** [comment]
Thanks, got cross-compilation working (not easy on Windows but it works
now).

First question - in persistent.go, what prevents a connection that is
currently in use from being returned t.Dial in connManager()?

**Comment by miekg** [comment]
There being removed from the list and then this conn is sent down the
channel

**Comment by WinstonPrivacy** [comment]
Got it. Thanks. Starting to investigate.

**Comment by WinstonPrivacy** [comment]
I have instrumented up persistent.go just to do some quick sanity checks
and something doesn't look right. I've added counters to Transport to
record the cumulative number of connection pool hits, misses, yields
(connections returned to the pool) and closed/expired connections.

The numbers don't add up. In theory, every cache miss should open a new TLS
connection. When that's done, it should be returned to the pool. However, I
consistently find that the # of cache misses is significantly higher than
the # of connections in the pool.

To confirm that connections aren't being closed, I set the t.expire time to
a very long time.Duration so connections are never expired and closed.

For instance, in the yield code, here's a line of my logging code. It shows
that at this time, there were 116 cumulative cache (connection pool) hits,
119 yields, and 3 connections in the pool. Those numbers add up. But it
also shows that we had 17 cache misses, each of which should open a new
connection.

[DEBUG] Persistent.go:connManager():yield addr 1.0.0.1:853 hits:116
misses:17 yields:119 closed 0 # connections in pool 3

Is there any reason a cache miss would not open a new connection? If not,
then there it looks like some unnecessary TLS handshakes are happening in
the background.

**Comment by rdrozhdzh** [comment]
@WinstonPrivacy, in case if write or read operation is failed (e.g. because of timeout) the connection is closed and not yielded to connection manager, see
https://github.com/coredns/coredns/blob/a6dbc837b3f04d0acda790fbccadf127945c4164/plugin/forward/connect.go#L97-L123

**Comment by WinstonPrivacy** [comment]
I'll take a closer look but this still doesn't seem right. I'm currently
seeing that 25-30% of TLS connections are erroring out and not being added
to the pool. The connection pool rarely goes above 2 warm connections and
new TLS connections are frequently being dialed.

**Comment by miekg** [comment]
I've seen that low level of connections as well, with normal UDP sockets.
The exported metric said there were only 3 (server was doing 30kqps), but 3
seemed low. Haven't made the time to look deeper into this, but obviously
still curious

**Comment by WinstonPrivacy** [comment]
I'm trying to wrap my head around it now.

I also see that CoreDNS is automatically recovering from panics. There are
some being triggered in the Forward plugin, specifically in
connect.go:Dial() ("runtime error: invalid memory address or nil pointer
dereference"). The ChicagoTribune.com just triggered it, which
coincidentally is the same site that was running CPU up to 100% in earlier
tests.

I'd like to narrow this down some more. Could it be related to the
unprotected hash map access?

**Comment by miekg** [comment]
Ha! Nice we getting somewhere. You arch might be triggering this sooner,
but this is good.

We had discussion on that map access, and deemed it safe, but apparently
not. I don't think a mutex adds much overhead here.

debug plugin should disable panic recovery btw

**Comment by miekg** [comment]
Specially the clean routine never sat well with me, but I failed to break
it or see a direct flaw

**Comment by WinstonPrivacy** [comment]
Great tip on the debug plugin. Trying to reproduce now but it is very
intermittent.

**Comment by WinstonPrivacy** [comment]
Hash map access appears to be safe (at least I haven't been able to trigger
any errors). The panics were coming from my debug code, but that did lead
to this - I am seeing large bursts of errors coming back from
connect.go:Dial() from dns.DialTimeoutWithTLS().

[image: image.png]

**Comment by miekg** [comment]
Is this because we dynamicly set timeouts based on UDP and apply those to
TLS TCP conns as well.

Could to know map access is ok

**Comment by WinstonPrivacy** [comment]
Getting closer.

The connection pool appears to properly fill up with warm connections after
a burst of traffic. Then it mysteriously drops all of them. The cleanup()
method is not being called when this happens. Rather, connect.go:Connect()
is erroring out here:

```

ret, err = conn.ReadMsg()
if err != nil {
       fmt.Println("[ERROR] err while getting response from
connection", err, "cached?", cached)
       conn.Close() // not giving it back
       if err == io.EOF && cached {
              return nil, ErrCachedClosed
       }
       return ret, err
}

 ```

This looks like this on the command line:

[image: image.png]

**Comment by WinstonPrivacy** [comment]
Trying the image again...

![image](https://user-images.githubusercontent.com/36650369/55821575-6f10b800-5ac3-11e9-9157-243bfa254261.png)

**Comment by rdrozhdzh** [comment]
this means remote server closed these connections

**Comment by miekg** [comment]
Yeah. Perfectly normal to close conns after some while.

**Comment by WinstonPrivacy** [comment]
Got it, that makes sense. I've replaced that error message with debug output to indicate when this is happening. This explains some of the discrepancy but connections are still being dropped in some manner that I'm not seeing.

In the screenshot below, we have 9 connections in the pool. The remote server closes six of them. We should have three left but the pool is empty afterward.

![image](https://user-images.githubusercontent.com/36650369/55822831-0e36af00-5ac6-11e9-9048-756aaead8361.png)

**Comment by rdrozhdzh** [comment]
in last line `hits 153` - this is greater by 9 than previous record. I.e. 9 connections were taken from cache, 6 of them were stale, 3 other - unknown, probably closed by other error, or still in use.

**Comment by WinstonPrivacy** [comment]
Ok. We're down to about 3% of connections being unaccounted for at this point so I'll let it go. 

I also noticed that we were seeing a fair number of broken pipe errors. Looks like this may have been related to our setting the default expires time to 90 seconds... remote servers were breaking connections before then. I reduced it to 30 seconds and I'm seeing fewer broken pipes now.

Moving on to the stress test.

**Comment by WinstonPrivacy** [comment]
Latest version recovers from a DNS stress test more quickly than before - ~45 seconds. I have not gotten it to destabilize yet with the following Corefile:

```
.:53 {
debug
       chaos Winston-DNS-1.0 support@winstonprivacy.com
       cache 3600
       hosts /etc/winston/hosts {
               fallthrough
       }

       forward . tls://1.1.1.1 {
               tls_servername cloudflare-dns.com
               health_check 5s
           max_fails 1
               expire 30s
       }
}
```

I then added a pending request counter to Connect() which automatically fails connections if we exceed 50 pending requests. With this in place, CoreDNS recovers in about 9 seconds. I am also able to (mostly) connect to new sites during the stress test, which was never possible before. Still getting a lot of tls.DialWithDialer timeouts and the connection pool is stuck at 0 free connections (maybe that makes sense because they are all in use). A big improvement overall.

**Comment by WinstonPrivacy** [comment]
We're going to start testing this more thoroughly shortly but so far it is performing well in automated stress tests. I added an optional "max_requests" stanza to the Forward block. If set, it will drop any requests if there are already this number of pending requests in progress. In practice, the limit is not a hard one and I'm seeing up to double this number in particularly strenuous test conditions.

I would suggest that this setting is probably only needed on low-end resource constrained boxes as CoreDNS seems to do well in most conditions.

I have not been able to destabliize CoreDNS after adding this patch. Under heavy load, it recovers and CPU drops back down to ~0% in 5-10 seconds after the load is discontinued.

https://github.com/WinstonPrivacyInc/coredns/blob/master/plugin/forward/connect.go

This is the relevant code (note that other small changes were needed to parse the Forward block):

```
numpending := atomic.AddInt64(&p.transport.pendingrequests, 1)
	defer func() {
		atomic.AddInt64(&p.transport.pendingrequests, -1)
	}()

	if p.transport.maxrequests > 0 && numpending > p.transport.maxrequests {
		return nil, fmt.Errorf("Too busy: %d requests pending", numpending)
}
```

**Comment by miekg** [comment]
Nice.

Is there any way you can calculate that value of 50 from the number of
CPUs? Otherwise having such a setting might make sense, although I'm still
allergic to magic numbers and user knobs

**Comment by WinstonPrivacy** [comment]
We could easily say 25x # of cores but that's really just a guess. While I
appreciate the auto-tuning philosophy, I have no idea what the impact might
be on other hardware or software deployments and wouldn't recommend anyone
use this unless they really needed it (which might be the case on low-end
ARM boards).

**Comment by miekg** [comment]
# Re: num-pending

this `return nil, fmt.Errorf("Too busy: %d requests pending", numpending)` does this mean the forward-resolve of this query is being terminated, i.e. an error is returned to the client? I think a better way is to let it try to get a `Conn` for a bit and only after the 5s timeout (default in *forward* anyway), return an error.

If you error to a client, it will probably retry *anyway*.

# Re: knobs and options

This make sense. But for any knob you need a) a good default b) advice on how to tune it (i.e. what metrics should I look at to make an informed decision on this value?

**Comment by WinstonPrivacy** [comment]
That's an excellent suggestion. We implement throttling elsewhere for
similar reasons. I would recommend not getting a connection but simply
sleeping for 200-400ms to slow down callers (you could potentially try
again but as you say, clients can be counted on to do this). On our
hardware, if there are 50-100 pending requests already in flight, then the
chances of the request being completed is near nil anyway.

I defaulted max_requests to 0, on the basis that it should not affect
anyone who doesn't specifically need it. How I determine the optimal value
was to perform a dnsperf stress test while counting the # of pending
requests. We were steadily climbing to 500 and the box was completely
unusable. By lowering it to 50, CoreDNS is actually capable of withstanding
a DOS/stress test while other processes continue to function properly.

**Comment by miekg** [comment]
@WinstonPrivacy I'm assuming your excellent sleuthing will result in a PR (soon-ish?)

Thanks!

**Comment by JasonSwindle** [comment]
@WinstonPrivacy @willwp    Any update?
